File: examples/ruby/spec/elements/locators_spec.rb
# frozen_string_literal: true

require 'spec_helper'

RSpec.describe 'Element Locators', skip: 'These are reference following the documentation example' do
  it 'finds element by class name' do
    driver.find_element(class: 'information')
  end

  it 'finds element by css selector' do
    driver.find_element(css: '#fname')
  end

  it 'finds element by id' do
    driver.find_element(id: 'lname')
  end

  it 'find element by name' do
    driver.find_element(name: 'newsletter')
  end

  it 'finds element by link text' do
    driver.find_element(link_text: 'Selenium Official Page')
  end

  it 'finds element by partial link text' do
    driver.find_element(partial_link_text: 'Official Page')
  end

  it 'finds element by tag name' do
    driver.find_element(tag_name: 'a')
  end

  it 'finds element by xpath' do
    driver.find_element(xpath: "//input[@value='f']")
  end

  context 'with relative locators' do
    it 'finds element above' do
      driver.find_element({relative: {tag_name: 'input', above: {id: 'password'}}})
    end

    it 'finds element below' do
      driver.find_element({relative: {tag_name: 'input', below: {id: 'email'}}})
    end

    it 'finds element to the left' do
      driver.find_element({relative: {tag_name: 'button', left: {id: 'submit'}}})
    end

    it 'finds element to the right' do
      driver.find_element({relative: {tag_name: 'button', right: {id: 'cancel'}}})
    end

    it 'finds near element' do
      driver.find_element({relative: {tag_name: 'input', near: {id: 'lbl-email'}}})
    end

    it 'chains relative locators' do
      driver.find_element({relative: {tag_name: 'button', below: {id: 'email'}, right: {id: 'cancel'}}})
    end
  end
end

File: examples/javascript/test/elements/locators.spec.js
const {By, Builder} = require('selenium-webdriver');
const assert = require("assert");

describe('Element Locator Test', function () {
  it('Check if element can be found by class name', async () => {
    let driver = await new Builder().forBrowser('chrome').build();
    await driver.get('https://www.selenium.dev/selenium/web/locators_tests/locators.html');
    const element = await driver.findElement(By.className('information'));
    const tag = await element.getTagName();
    const id = await element.getAttribute("id");
    const value = await element.getAttribute("value");

    assert.equal(tag, "input");
    assert.equal(id, "fname");
    assert.equal(value, "Jane");
    await driver.quit();
  });

  it('Check if element can be found by css selector', async function () {
    let driver = await new Builder().forBrowser('chrome').build();
    await driver.get('https://www.selenium.dev/selenium/web/locators_tests/locators.html');
    const element = await driver.findElement(By.css('#fname'));

    const tag = await element.getTagName();
    const id = await element.getAttribute("id");
    const value = await element.getAttribute("value");

    assert.equal(tag, "input");
    assert.equal(id, "fname");
    assert.equal(value, "Jane");
    await driver.quit();
  });

  it('Check if element can be found by id', async function () {
    let driver = await new Builder().forBrowser('chrome').build();
    await driver.get('https://www.selenium.dev/selenium/web/locators_tests/locators.html');
    const element = await driver.findElement(By.id('lname'));

    const tag = await element.getTagName();
    const id = await element.getAttribute("id");
    const value = await element.getAttribute("value");

    assert.equal(tag, "input");
    assert.equal(id, "lname");
    assert.equal(value, "Doe");
    await driver.quit();
  });

  it('Check if element can be found by name', async function () {
    let driver = await new Builder().forBrowser('chrome').build();
    await driver.get('https://www.selenium.dev/selenium/web/locators_tests/locators.html');
    const element = await driver.findElement(By.name("newsletter"));

    const tag = await element.getTagName();
    const type = await element.getAttribute("type");
    const value = await element.getAttribute("value");

    assert.equal(tag, "input");
    assert.equal(type, "checkbox");
    assert.equal(value, "1");
    await driver.quit();
  });



  it('Check if element can be found by link text', async function () {
    let driver = await new Builder().forBrowser('chrome').build();
    await driver.get('https://www.selenium.dev/selenium/web/locators_tests/locators.html');
    const element = await driver.findElement(By.linkText("Selenium Official Page"));

    const tag = await element.getTagName();
    const href = await element.getAttribute("href");

    assert.equal(tag, "a");
    assert.equal(href, "https://www.selenium.dev/");
    await driver.quit();
  });

  it('Check if element can be found by partial link text', async function () {
    let driver = await new Builder().forBrowser('chrome').build();
    await driver.get('https://www.selenium.dev/selenium/web/locators_tests/locators.html');
    const element = await driver.findElement(By.partialLinkText("Official Page"));

    const tag = await element.getTagName();
    const href = await element.getAttribute("href");

    assert.equal(tag, "a");
    assert.equal(href, "https://www.selenium.dev/");
    await driver.quit();
  });

  it('Check if element can be found by tag name', async function () {
    let driver = await new Builder().forBrowser('chrome').build();
    await driver.get('https://www.selenium.dev/selenium/web/locators_tests/locators.html');
    const element = await driver.findElement(By.tagName("a"));

    const tag = await element.getTagName();
    const text = await element.getText();

    assert.equal(tag, "a");
    assert.equal(text, "Selenium Official Page");
    await driver.quit();
  });

  it('Check if element can be found by xpath', async function () {
    let driver = await new Builder().forBrowser('chrome').build();
    await driver.get('https://www.selenium.dev/selenium/web/locators_tests/locators.html');
    const element = await driver.findElement(By.xpath('//input[@value="f"]'));

    const tag = await element.getTagName();
    const type = await element.getAttribute("type");
    const value = await element.getAttribute("value");

    assert.equal(tag, "input");
    assert.equal(type, "radio");
    assert.equal(value, "f");
    await driver.quit();
  });
});
